controller/setting_controller.go

```diff
@@ -606,9 +588,19 @@ func (sc *SettingController) updateTaintToleration() error {
 		if err != nil {
 			return err
 		}
+
 		if reflect.DeepEqual(util.TolerationListToMap(lastAppliedTolerationsList), newTolerationsMap) {
 			continue
 		}
+
+		detached, _, err := sc.ds.AreAllVolumesDetached(longhorn.DataEngineTypeAll)
+		if err != nil {
+			return errors.Wrapf(err, "failed to check volume detachment for %v setting update", types.SettingNameTaintToleration)
+		}
+		if !detached {
+			return &types.ErrorInvalidState{Reason: fmt.Sprintf("failed to apply %v setting to Longhorn components when there are attached volumes. It will be eventually applied", types.SettingNameTaintToleration)}
+		}
+
```

Comment:
Can we relocate the check just before the for loop rather than inside the loop?

Reply:
Agree that we should check it once before the loop. WDYT about something like this:
```go
func hasUpdatedToleration(newTolerationsMap map[string]corev1.Toleration, objs ...runtime.Object) (bool, error) {
	for _, obj := range objs {
		lastAppliedTolerationsList, err := getLastAppliedTolerationsList(obj)
		if err != nil {
			return false, err
		}
		if !reflect.DeepEqual(util.TolerationListToMap(lastAppliedTolerationsList), newTolerationsMap) {
			return true, nil
		}
	}

	return false, nil
}

// updateTaintToleration deletes all user-deployed and system-managed components immediately with the updated taint toleration.
func (sc *SettingController) updateTaintToleration() error {
	setting, err := sc.ds.GetSettingWithAutoFillingRO(types.SettingNameTaintToleration)
	if err != nil {
		return err
	}
	newTolerations := setting.Value
	newTolerationsList, err := types.UnmarshalTolerations(newTolerations)
	if err != nil {
		return err
	}
	newTolerationsMap := util.TolerationListToMap(newTolerationsList)
	runtimeObjects := []runtime.Object

	daemonsetList, err := sc.ds.ListDaemonSetWithLabels(types.GetBaseLabelsForSystemManagedComponent())
	if err != nil {
		return errors.Wrap(err, "failed to list Longhorn daemonsets for toleration update")
	}
	for _, obj := range daemonsetList {
		runtimeObjects = append(runtimeObjects, runtime.Object(obj))
	}

	deploymentList, err := sc.ds.ListDeploymentWithLabels(types.GetBaseLabelsForSystemManagedComponent())
	if err != nil {
		return errors.Wrap(err, "failed to list Longhorn deployments for toleration update")
	}
	for _, obj := range deploymentList {
		runtimeObjects = append(runtimeObjects, runtime.Object(obj))
	}

	imPodList, err := sc.ds.ListInstanceManagerPods()
	if err != nil {
		return errors.Wrap(err, "failed to list instance manager pods for toleration update")
	}
	for _, obj := range imPodList {
		runtimeObjects = append(runtimeObjects, runtime.Object(obj))
	}

	smPodList, err := sc.ds.ListShareManagerPods()
	if err != nil {
		return errors.Wrap(err, "failed to list share manager pods for toleration update")
	}
	for _, obj := range smPodList {
		runtimeObjects = append(runtimeObjects, runtime.Object(obj))
	}

	bimPodList, err := sc.ds.ListBackingImageManagerPods()
	if err != nil {
		return errors.Wrap(err, "failed to list backing image manager pods for toleration update")
	}
	for _, obj := range bimPodList {
		runtimeObjects = append(runtimeObjects, runtime.Object(obj))
	}


	hasUpdated, err := hasUpdatedToleration(newTolerationsMap, runtimeObjects...)
	if err != nil {
		return err
	}
	if hasUpdated {
		return nil
	}
```

Reply:
@derekbit 
Moved the whole section out of the loops and relocated the check volume attached section after checking the setting synchronized.

@PhanLe1010
```golang
func hasUpdatedToleration(newTolerationsMap map[string]corev1.Toleration, objs ...runtime.Object) (bool, error) {
	for _, obj := range objs {
		lastAppliedTolerationsList, err := getLastAppliedTolerationsList(obj)
		if err != nil {
			return false, err
		}
		if !reflect.DeepEqual(util.TolerationListToMap(lastAppliedTolerationsList), newTolerationsMap) {
			return true, nil
		}
	}

	return false, nil
}
...
```
There is a scenario: 
1. If we have three setting controllers on three worker nodes,
- controller A starts updating deployment A >B >C
- controller B starts updating deployment B >A >C
- controller C starts updating deployment B >A >C

1. When controller A tries to update B, it will get a conflict error and then be re-queued and reconcile the setting again (the situation as controller B and C.)
1. When controllers A, B, and C start to reconcile the setting again, it will get deployment A, and B updated and return nil as the setting synchronized but deployment C does not apply the setting.

If we check which objects are not synchronized and return an unsynchronized list at the end of  `hasUpdatedToleration`, it should be able to handle the scenario above.

The implementation was updated as you suggested.